WOMBATlite-sinking: Spatially variable sinking rates within the generic tracers version of WOMBAT-lite

pearseb opened the pull request:
To include variable sinking rates in WOMBAT-lite within the generic tracers framework I have had to slightly extend the tri-diagonal (implicit) solver for vertical diffusion within the _generic_tracer_utils.F90_ file. Specifically, I have altered the "IOWtridiag" algorithm in the _g_tracer_vertdiff_M_ subroutine. This algorithm previously solved only for the vertical movement of tracer within the water column, but did not account for the passing of tracer from the ocean model into a bottom reservoir. I have simply extended this algorithm to include the passing of some tracer into a bottom reservoir (i.e., sediment pool).

I have also made additions in the form of:

* Zooplankton grazing on organic detritus
* More diagnostics of zooplankton grazing, sloppy feeding and excretion associated with both phytoplankton and detritus prey, as well as a remineralisation rate diagnostic
* Slight alterations to the parameters (not final)
* Updates to nutrient limitation calculations, including the fixing of a bug in the iron limitation term
* More diagnostics related to iron cycling
* Movement of DIC and aDIC tracers into the sources and sinks nested timestep as part of the ecosystem cycling
* Removal of virtual fluxes which was causing unrealistic accumulation of NO3, DIC, Alk in the surface ocean.

dougiesquire reviewed (commented):
Thanks @pearseb. A couple of very minor comments/questions from an initial pass. As a general rule, I don't think we want to include commented out lines of code, especially if there is no comment explaining why they are commented out.

Regarding the vmove changes in `generic_tracer_utils.F90`, could we chat in person? I think we should definitely try to get these changes upstream into the main generic tracers repo and I have a few other things to discuss. I'll message you on Slack to sort out a time.

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -2651,6 +2825,33 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
+    !-----------------------------------------------------------------------
+    ! Assign spatially variable vertical movement of tracers
+    !-----------------------------------------------------------------------
+    call g_tracer_get_pointer(tracer_list, 'det', 'vmove', wombat%p_wdet) ! [mol/kg]
+    call g_tracer_get_pointer(tracer_list, 'detfe', 'vmove', wombat%p_wdetfe) ! [mol/kg]
```
Yep they certainly do. Changed.

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
Agreed. Sorry. Removed.

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -2431,8 +2560,9 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
-        if (wombat%zw(i,j,k) .ge. 180.0) wombat%detremi(i,j,k) = 0.5 * wombat%detremi(i,j,k) ! reduce decay below 180m
+        wombat%detremi(i,j,k) = wombat%reminr(i,j,k) / mmol_m3_to_mol_kg * &
+                                wombat%f_det(i,j,k) * wombat%f_det(i,j,k) ! [molC/kg/s]
+      !  if (wombat%zw(i,j,k) .ge. 180.0) wombat%detremi(i,j,k) = 0.5 * wombat%detremi(i,j,k) ! reduce decay below 180m
```
Yep. Done.

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -2125,49 +2236,46 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
 
       ! Calculate impact of euphotic zone depth on phytoplankton and chlorophyll production
-      wombat%phy_leup(i,j) = MIN(1.0, wombat%zeuphot(i,j)/(hblt_depth(i,j) + epsi))
+!      wombat%phy_leup(i,j) = MIN(1.0, wombat%zeuphot(i,j)/(hblt_depth(i,j) + epsi))
+       wombat%phy_leup(i,j) = 1.0
```
Removed.

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -1396,6 +1503,7 @@ subroutine user_add_tracers(tracer_list)
         units = 'mol/kg', &
         prog = .true., &
         sink_rate = wombat%wcaco3, &
+        move_vertical = .false., &
```
Done.

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -1385,7 +1490,9 @@ subroutine user_add_tracers(tracer_list)
         units = 'mol/kg', &
         prog = .true., &
-        sink_rate = wombat%wdetbio, &
+        sink_rate = wombat%wdetbio*0.0, &
```
Done.

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -1375,7 +1478,9 @@ subroutine user_add_tracers(tracer_list)
         units = 'mol/kg', &
         prog = .true., &
-        sink_rate = wombat%wdetbio, &
+        sink_rate = wombat%wdetbio*0.0, &
```
Done.

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -1228,7 +1330,8 @@ subroutine user_add_params
     !-----------------------------------------------------------------------
-    call g_tracer_add_param('no3_global', wombat%no3_global, 4.512e-06)
+    !call g_tracer_add_param('no3_global', wombat%no3_global, 4.512e-06)
+    call g_tracer_add_param('no3_global', wombat%no3_global, 0.0)
```
Yes it was temporary. Now removed.

pearseb commented:
Hi @dougiesquire,
I've addressed your suggestions above. Sorry for the annoying comments in the code.
I'm now just ensuring that the budgets all close, so there may be one or two additional commits to go until I'm satisfied.
Pearse

dougiesquire commented on generic_tracers/generic_tracer_utils.F90:
```diff
@@ -3181,8 +3181,10 @@ subroutine g_tracer_vertdiff_G(g_tracer, h_old, ea, eb, dt, kg_m2_to_H, m_to_H,
-	    do k=2,nz; sink_dist(k) = (dt*g_tracer%vmove(i,j,k)) * m_to_H; enddo
-	  endif
+            do k=2,nz; sink_dist(k) = (dt*g_tracer%vmove(i,j,k)) * m_to_H; enddo
+            sink_dist(nz+1) = (dt*g_tracer%vmove(i,j,nz)) * m_to_H !PJB [13th Nov 2024] to allow sinking into bottom reservoir
+            ! PJB... although this doesn't work because "move_vertical==.true." makes "GOLDtridiag==.false."
```
As discussed, let's keep this in since this code is used by MOM6. But I think we could remove these lines and do the same thing more cleanly by changing 3196 from:
```
sink(nz+1) = sink_dist(nz+1)
```
to
```
sink(nz+1) = sink_dist(nz)
```

pearseb replied:
Agreed and changed. Looks to work well.

dougiesquire replied:
@pearseb I don't think you've pushed your changes?

pearseb replied:
Changes pushed now

dougiesquire commented on generic_tracers/generic_tracer_utils.F90:
```diff
@@ -3316,8 +3318,8 @@ subroutine g_tracer_vertdiff_M(g_tracer,dh, dhw, diff_cbt, dt, rho0,tau)
     GOLDtridiag = .true.
     IOWtridiag  = .false.
     if (g_tracer%move_vertical) then
-       GOLDtridiag = .false.
-       IOWtridiag  = .true.
+      GOLDtridiag = .false.
+      IOWtridiag  = .true.
```
As discussed, let's undo all the syntactical changes to this file

pearseb replied:
This should be corrected.

dougiesquire replied:
@pearseb I don't think you've pushed your changes?

pearseb replied:
Changes pushed now

dougiesquire reviewed (commented):
A couple more minor comments, now on the `generic_tracer_utils.F90` changes. As we discussed, once you've addressed these we can rebase the changes to `generic_tracer_utils.F90` into a single commit and open a PR upstream.

dougiesquire commented:
@pearseb, I don't think you've pushed all your changes. `generic_tracer_utils.F90` hasn't changed since my last review

pearseb commented:
Hi Dougie,
Sorry it took so much extra time. I was getting some very annoying problems with the air-sea flux of DIC. Seemingly without reason, every now and then the model would stop because of a massive influx of DIC near the regions of river input to the ocean. The DIC tracer would spike to ridiculous concentrations as a result. These instabilities seem to be avoided by placing upper and lower limits on the DIC concentrations that are read into the ocmip2_co2calc subroutine. As far as I can tell, there seems to be no problems with this fix. I was worried that by setting upper and lower limits, that the DIC tracer might constantly bleed large or small concentrations into the wider ocean from these points. But this doesn't seem to be the case, as the spikes in concentration that affect the air-sea fluxes occur for one timestep only, and afterwards disappear.
Pearse

dougiesquire commented on generic_tracers/generic_tracer_utils.F90:
```diff
@@ -3231,6 +3228,11 @@ subroutine g_tracer_vertdiff_G(g_tracer, h_old, ea, eb, dt, kg_m2_to_H, m_to_H,
              c1(k) = eb(i,j,k-1) * b1
              b_denom_1 = h_minus_dsink(k) + d1 * (ea(i,j,k) + sink(k))
+             if (b_denom_1.le.0.0) then
+               print*, "Warning: b_denom_1 <= 0 at (i, j, k):", i, j, k
+               print*, "Values: h_minus_dsink =", h_minus_dsink(k), "d1 =", d1, "ea =", ea(i,j,k), "sink =", sink(k)
+               stop "Error: Diagonal value b_denom_1 must be > 0.0"
```
We should use FMS to log the error here

```suggestion
               call mpp_error(FATAL, trim(sub_name)//": Diagonal value b_denom_1 must be > 0.0")
```

Will also need to add the following to the start of the routine:

```
character(len=fm_string_len), parameter :: sub_name = 'g_tracer_vertdiff_G'
```

dougiesquire commented on generic_tracers/generic_tracer_utils.F90:
```diff
@@ -3315,10 +3317,11 @@ subroutine g_tracer_vertdiff_M(g_tracer,dh, dhw, diff_cbt, dt, rho0,tau)
-       GOLDtridiag = .false.
-       IOWtridiag  = .true.
-    endif
+    ! PJB [1st Dec 2024] Commenting out the next 4 lines ensures GOLDtridiag even when move_vertical==.true.
```
```suggestion
    ! PJB [1st Dec 2024] Commenting out since GOLDtridiag now works with move_vertical==.true.
```

dougiesquire commented on generic_tracers/generic_tracer_utils.F90:
```diff
@@ -3415,34 +3418,41 @@ subroutine g_tracer_vertdiff_M(g_tracer,dh, dhw, diff_cbt, dt, rho0,tau)
                 fact1  = dt/dh(i,j,k)
-		fact2  = rho0*fact1*0.5
-		factu  = fact1/dhw(i,j,km1)
+       fact2  = rho0*fact1*0.5
+       factu  = fact1/dhw(i,j,km1)
```
```suggestion
		fact2  = rho0*fact1*0.5
		factu  = fact1/dhw(i,j,km1)
```

dougiesquire commented on generic_tracers/generic_tracer_utils.F90:
```diff
@@ -3415,34 +3418,41 @@ subroutine g_tracer_vertdiff_M(g_tracer,dh, dhw, diff_cbt, dt, rho0,tau)
-		wnegl(i,k)  = fact2*(g_tracer%vmove(i,j,k  ) - wabsl)*g_tracer_com%grid_tmask(i,j,kp1)
-                a1(i,k) = dcb(i,j,km1)*factu*g_tracer_com%grid_tmask(i,j,k)  
-                c1(i,k) = dcb(i,j,k)  *factl*g_tracer_com%grid_tmask(i,j,kp1)
+       wabsu       = abs(g_tracer%vmove(i,j,km1))
+       wposu(i,k)  = fact2*(g_tracer%vmove(i,j,km1) + wabsu)*g_tracer_com%grid_tmask(i,j,k)
+       wnegu(i,k)  = fact2*(g_tracer%vmove(i,j,km1) - wabsu)*g_tracer_com%grid_tmask(i,j,k)
+       wabsl       = abs(g_tracer%vmove(i,j,k))
+       wposl(i,k)  = fact2*(g_tracer%vmove(i,j,k  ) + wabsl)*g_tracer_com%grid_tmask(i,j,kp1)
+       wnegl(i,k)  = fact2*(g_tracer%vmove(i,j,k  ) - wabsl)*g_tracer_com%grid_tmask(i,j,kp1)
```
```suggestion
		wabsu       = abs(g_tracer%vmove(i,j,km1))
		wposu(i,k)  = fact2*(g_tracer%vmove(i,j,km1) + wabsu)*g_tracer_com%grid_tmask(i,j,k)
		wnegu(i,k)  = fact2*(g_tracer%vmove(i,j,km1) - wabsu)*g_tracer_com%grid_tmask(i,j,k)
		wabsl       = abs(g_tracer%vmove(i,j,k))
		wposl(i,k)  = fact2*(g_tracer%vmove(i,j,k  ) + wabsl)*g_tracer_com%grid_tmask(i,j,kp1)
		wnegl(i,k)  = fact2*(g_tracer%vmove(i,j,k  ) - wabsl)*g_tracer_com%grid_tmask(i,j,kp1)
```

dougiesquire commented on generic_tracers/generic_tracer_utils.F90:
```diff
@@ -3415,34 +3418,41 @@ subroutine g_tracer_vertdiff_M(g_tracer,dh, dhw, diff_cbt, dt, rho0,tau)
+                  g_tracer%btm_reservoir(i,j) = g_tracer%btm_reservoir(i,j) - &
+                                                wnegl(i,k)*g_tracer%field(i,j,k,tau)*dh(i,j,k)
+                endif
+                a1(i,k) = dcb(i,j,km1)*factu*g_tracer_com%grid_tmask(i,j,k) 
+                c1(i,k) = dcb(i,j,k)  *factl*g_tracer_com%grid_tmask(i,j,kp1) 
```
Changing back to original

```suggestion
                a1(i,k) = dcb(i,j,km1)*factu*g_tracer_com%grid_tmask(i,j,k)  
                c1(i,k) = dcb(i,j,k)  *factl*g_tracer_com%grid_tmask(i,j,kp1)
```

dougiesquire commented on generic_tracers/generic_tracer_utils.F90:
```diff
@@ -3415,34 +3418,41 @@ subroutine g_tracer_vertdiff_M(g_tracer,dh, dhw, diff_cbt, dt, rho0,tau)
              a1(i,1)  = 0.0
-	     wnegu(i,1) = 0.0; wposu(i,1) = 0.0
-	     a(i,1)  = 0.0
+        wnegu(i,1) = 0.0; wposu(i,1) = 0.0
+        a(i,1)  = 0.0
```
```suggestion
	     wnegu(i,1) = 0.0; wposu(i,1) = 0.0
	     a(i,1)  = 0.0
```

dougiesquire commented on generic_tracers/generic_tracer_utils.F90:
```diff
@@ -3415,34 +3418,41 @@ subroutine g_tracer_vertdiff_M(g_tracer,dh, dhw, diff_cbt, dt, rho0,tau)
+        a(i,1)  = 0.0
              c1(i,g_tracer_com%nk) = 0.0
-	     wposl(i,g_tracer_com%nk) = 0.0; wnegl(i,g_tracer_com%nk) = 0.0 
+        wposl(i,g_tracer_com%nk) = 0.0
```
```suggestion
             wposl(i,g_tracer_com%nk) = 0.0
```

dougiesquire commented on generic_tracers/generic_tracer_utils.F90:
```diff
@@ -3415,34 +3418,41 @@ subroutine g_tracer_vertdiff_M(g_tracer,dh, dhw, diff_cbt, dt, rho0,tau)
              b(i,1)  = 1.0 + a1(i,1) + c1(i,1) - wnegl(i,1) + wposu(i,1)
              b(i,g_tracer_com%nk) = 1.0 + a1(i,g_tracer_com%nk) + c1(i,g_tracer_com%nk) &
-	                                - wnegl(i,g_tracer_com%nk) + wposu(i,g_tracer_com%nk)
+                                   - wnegl(i,g_tracer_com%nk) + wposu(i,g_tracer_com%nk)
```
```suggestion
	                                - wnegl(i,g_tracer_com%nk) + wposu(i,g_tracer_com%nk)
```

dougiesquire reviewed (commented):
Thanks @pearseb. I've added some initial comments and suggestions on `FMS_ocmip2_co2calc.F90` and `generic_tracer_utils.F90`. I'll come by your office to discuss.

Let's chat through these then I'll dive into `generic_WOMBATlite.F90`

dougiesquire reviewed (commented):
Some more comments, now on `generic_WOMBATlite.F90`. I'm curious, have you tried to run this with OM3? Also have you got a feel for the performance impacts of all these changes?

pearseb commented on generic_tracers/FMS_ocmip2_co2calc.F90:
```diff
@@ -209,7 +209,7 @@ subroutine FMS_ocmip2_co2calc(dope_vec, mask,                      &
 !       local variables
 !
 integer :: isc, iec, jsc, jec
-integer :: i,j
```
We can just use mocsy but I think that I would like to maintain the changes to this file because it provides 3D output from the "ocmip2" option.

dougiesquire replied:
In https://github.com/ACCESS-NRI/GFDL-generic-tracers/pull/15/commits/37f201b47cf84fb98aaebb68c33973a67f9b8bbb I added an optional parameter, `pcorr_ocmip2`, that let's users preserve answers

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -1147,30 +1355,70 @@ subroutine user_add_params
+    !-----------------------------------------------------------------------
+    call g_tracer_add_param('wcaco3', wombat%wcaco3, 4.0/86400.0)
+    
+    ! Do dynamic CaCO3 precipiation, dissolution and ballasting? 
+    !-----------------------------------------------------------------------
+    call g_tracer_add_param('caco3_dynamics', wombat%caco3_dynamics, .true. )
+
+    ! Permanental burial organics and CaCO3 in sediments? 
+    !-----------------------------------------------------------------------
+    call g_tracer_add_param('burial', wombat%burial, .true. )
+
+    ! Add back the lost NO3 and Alk due to burial to surface? 
+    !-----------------------------------------------------------------------
+    call g_tracer_add_param('conservetracers', wombat%conservetracers, .true. )
```
Happy for this to be the case

dougiesquire replied:
Done in https://github.com/ACCESS-NRI/GFDL-generic-tracers/pull/15/commits/ebeb483ed932515728fca783b7b8a6a52eb77328

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -1413,23 +1663,33 @@ subroutine user_add_tracers(tracer_list)
+        longname = 'remineralised Dissolved Inorganic Carbon', &
+        units = 'mol/kg', &
+        prog = .true., &
+        flux_bottom = .true., &
+        flux_virtual = .true.)
```
Oh... You're right.

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -2125,80 +2440,71 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
 
       ! dts: in WOMBAT v3, export production was calucated before updating the source terms so we do
-      ! the same here
+      ! the same here. pjb: need to fix this.
```
Yeah so the problem is that export of material is now dependent on spatially variable sinking rates so I need to define this later. Will do so now.

dougiesquire replied:
Done in https://github.com/ACCESS-NRI/GFDL-generic-tracers/pull/15/commits/2ad0f96040fa3325d03308e1ad8d283177b23ab6

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
```diff
@@ -2222,30 +2528,31 @@ subroutine generic_WOMBATlite_update_from_source(tracer_list, Temp, Salt,  &
+      wombat%phy_kfe(i,j,k) = wombat%phykf * max(0.1, max(0.0, (biophy-wombat%phybiot))**0.37)
       ! Nitrogen limitation (currently Michaelis-Menten term)
-      wombat%phy_lnit(i,j,k) = biono3 / (biono3 + phy_k_nit + epsi)
+      wombat%phy_lnit(i,j,k) = biono3 / (biono3 + wombat%phy_kni(i,j,k))
```
It can't be unless the parameter "phykn" is set to 0.0 in the field_table. But yes we can add this back in it won't have any effect really since it will always be <<< biono3.

pearseb commented on generic_tracers/generic_WOMBATlite.F90:
I think that sounds great! Let's do that.

dougiesquire commented:
> This is looking good. @pearseb and I discussed that the last thing to do is a bit of git tidying to make some of the changes in this PR easier to submit upstream. The plan is to:
> 
>     * Make the extension to `FMS_ocmip2_co2calc.F90` in a separate PR and merge into `main`
> 
>     * Make the extension to `generic_tracer_utils.F90` in a separate PR and merge into `main`
> 
>     * Rebase this PR onto `main` (squash) and merge into `main`

This has now been done and we've confirmed that this rebased branch builds with ACCESS-OM2

dougiesquire reviewed (approved):
Looks good to me - sorry for taking so long to get to this!
